File: core/src/main/csharp/ch/cyberduck/core/collections/LRUCache.cs
// 
// Copyright (c) 2010 Yves Langisch. All rights reserved.
// http://cyberduck.ch/
// 
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// Bug fixes, suggestions and comments should be sent to:
// yves@cyberduck.ch
// 
using System;
using System.Collections;
using System.Collections.Generic;

namespace Ch.Cyberduck.Core.Collections
{
    public class LRUCache<TKey, TValue> : IDictionary<TKey, TValue>
    {
        private readonly int capacity;
        private readonly Dictionary<TKey, TValue> data;
        private readonly ICollection<KeyValuePair<TKey, TValue>> dataAsCollection;
        private readonly IndexedLinkedList<TKey> lruList = new IndexedLinkedList<TKey>();

        public LRUCache(int capacity)
        {
            if (capacity <= 0)
            {
                throw new ArgumentException("capacity should always be bigger than 0");
            }

            data = new Dictionary<TKey, TValue>(capacity);
            dataAsCollection = data;
            this.capacity = capacity;
        }

        public void Add(TKey key, TValue value)
        {
            if (!ContainsKey(key))
            {
                this[key] = value;
            }
            else
            {
                throw new ArgumentException("An attempt was made to insert a duplicate key in the cache.");
            }
        }

        public bool ContainsKey(TKey key)
        {
            return data.ContainsKey(key);
        }

        public ICollection<TKey> Keys
        {
            get { return data.Keys; }
        }

        public bool Remove(TKey key)
        {
            bool existed = data.Remove(key);
            lruList.Remove(key);
            return existed;
        }

        public bool TryGetValue(TKey key, out TValue value)
        {
            return data.TryGetValue(key, out value);
        }

        public ICollection<TValue> Values
        {
            get { return data.Values; }
        }

        public TValue this[TKey key]
        {
            get
            {
                var value = data[key];
                lruList.Remove(key);
                lruList.Add(key);
                return value;
            }
            set
            {
                data[key] = value;
                lruList.Remove(key);
                lruList.Add(key);

                if (data.Count > capacity)
                {
                    Remove(lruList.First);
                }
            }
        }

        public void Add(KeyValuePair<TKey, TValue> item)
        {
            Add(item.Key, item.Value);
        }

        public void Clear()
        {
            data.Clear();
            lruList.Clear();
        }

        public bool Contains(KeyValuePair<TKey, TValue> item)
        {
            return dataAsCollection.Contains(item);
        }

        public void CopyTo(KeyValuePair<TKey, TValue>[] array, int arrayIndex)
        {
            dataAsCollection.CopyTo(array, arrayIndex);
        }

        public int Count
        {
            get { return data.Count; }
        }

        public bool IsReadOnly
        {
            get { return false; }
        }

        public bool Remove(KeyValuePair<TKey, TValue> item)
        {
            bool removed = dataAsCollection.Remove(item);
            if (removed)
            {
                lruList.Remove(item.Key);
            }
            return removed;
        }

        public IEnumerator<KeyValuePair<TKey, TValue>> GetEnumerator()
        {
            return dataAsCollection.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return ((IEnumerable) data).GetEnumerator();
        }
    }
}

File: core/src/main/java/ch/cyberduck/core/cache/LRUCache.java
package ch.cyberduck.core.cache;

/*
 * Copyright (c) 2002-2018 iterate GmbH. All rights reserved.
 * https://cyberduck.io/
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.Collections;
import java.util.Map;
import java.util.concurrent.TimeUnit;
import java.util.function.Function;

import com.google.common.cache.Cache;
import com.google.common.cache.CacheBuilder;
import com.google.common.cache.CacheLoader;
import com.google.common.cache.LoadingCache;
import com.google.common.cache.RemovalListener;
import com.google.common.cache.RemovalNotification;
import com.google.common.util.concurrent.UncheckedExecutionException;

public class LRUCache<Key, Value> {
    private static final Logger log = LogManager.getLogger(LRUCache.class);

    public static <Key, Value> LRUCache<Key, Value> usingLoader(final Function<Key, Value> loader) {
        return usingLoader(loader, null, -1L, -1L);
    }

    public static <Key, Value> LRUCache<Key, Value> usingLoader(final Function<Key, Value> loader, final RemovalListener<Key, Value> listener) {
        return usingLoader(loader, listener, -1L, -1L);
    }

    public static <Key, Value> LRUCache<Key, Value> usingLoader(final Function<Key, Value> loader, final long maximumSize) {
        return usingLoader(loader, null, maximumSize, -1L);
    }

    public static <Key, Value> LRUCache<Key, Value> usingLoader(final Function<Key, Value> loader, final RemovalListener<Key, Value> listener, final long maximumSize) {
        return usingLoader(loader, listener, maximumSize, -1L);
    }

    public static <Key, Value> LRUCache<Key, Value> usingLoader(final Function<Key, Value> loader, final RemovalListener<Key, Value> listener, final long maximumSize, final long expireDuration) {
        return usingLoader(loader, listener, maximumSize, expireDuration, true);
    }

    public static <Key, Value> LRUCache<Key, Value> usingLoader(final Function<Key, Value> loader, final RemovalListener<Key, Value> listener, final long maximumSize, final long expireDuration, boolean expireAfterAccess) {
        return new LRUCache<>(loader, listener, maximumSize, expireDuration, expireAfterAccess);
    }

    public static <Key, Value> LRUCache<Key, Value> build() {
        return build(null, -1L, -1L);
    }

    public static <Key, Value> LRUCache<Key, Value> build(final RemovalListener<Key, Value> listener) {
        return build(listener, -1L, -1L);
    }

    public static <Key, Value> LRUCache<Key, Value> build(final long maximumSize) {
        return build(null, maximumSize, -1L);
    }

    public static <Key, Value> LRUCache<Key, Value> build(final RemovalListener<Key, Value> listener, final long maximumSize) {
        return build(listener, maximumSize, -1L);
    }

    public static <Key, Value> LRUCache<Key, Value> build(final RemovalListener<Key, Value> listener, final long maximumSize, final long expireDuration) {
        return build(listener, maximumSize, expireDuration, true);
    }

    public static <Key, Value> LRUCache<Key, Value> build(final RemovalListener<Key, Value> listener, final long maximumSize, final long expireDuration, boolean expireAfterAccess) {
        return new LRUCache<>(null, listener, maximumSize, expireDuration, expireAfterAccess);
    }

    private final Cache<Key, Value> delegate;

    private LRUCache(final Function<Key, Value> loader, final RemovalListener<Key, Value> listener, final long maximumSize, final long expireDuration, boolean expireAfterAccess) {
        final CacheBuilder<Object, Object> builder = CacheBuilder.newBuilder();
        if(listener != null) {
            builder.removalListener(new RemovalListener<Key, Value>() {
                @Override
                public void onRemoval(final RemovalNotification<Key, Value> notification) {
                    if(log.isDebugEnabled()) {
                        log.debug("Removed {} from cache with cause {}", notification.getKey(), notification.getCause());
                    }
                    listener.onRemoval(notification);
                }
            });
        }
        if(maximumSize > 0) {
            builder.maximumSize(maximumSize);
        }
        if(expireDuration > 0) {
            if(expireAfterAccess) {
                builder.expireAfterAccess(expireDuration, TimeUnit.MILLISECONDS);
            }
            else {
                builder.expireAfterWrite(expireDuration, TimeUnit.MILLISECONDS);
            }
        }
        if(loader != null) {
            delegate = builder.build(new CacheLoader<Key, Value>() {
                @Override
                public Value load(Key key) {
                    return loader.apply(key);
                }
            });
        }
        else {
            delegate = builder.build();
        }
    }

    public Value get(final Key key) throws UncheckedExecutionException {
        if(delegate instanceof LoadingCache) {
            return ((LoadingCache<Key, Value>) delegate).getUnchecked(key);
        }
        return delegate.getIfPresent(key);
    }

    public Map<Key, Value> asMap() {
        return Collections.unmodifiableMap(delegate.asMap());
    }

    public void put(final Key key, Value value) {
        if(null == key) {
            log.warn("Discard caching {}={}", key, value);
            return;
        }
        if(null == value) {
            this.remove(key);
        }
        else {
            delegate.put(key, value);
        }
    }

    public void remove(final Key key) {
        delegate.invalidate(key);
    }

    public long size() {
        return delegate.size();
    }

    public boolean isEmpty() {
        return delegate.size() == 0;
    }

    public boolean contains(final Key key) {
        return null != delegate.getIfPresent(key);
    }

    public void clear() {
        delegate.invalidateAll();
    }

    /**
     * Performs any pending maintenance operations needed by the cache
     */
    public void evict() {
        delegate.cleanUp();
    }
}
